### PR TITLE
nixos/caddy: fix overwriting of tls settings in config

### PR DESCRIPTION
**Pull request by lf-**
This fixes a bug I reported in the thread of #97217 in which jq was used to recursively merge the nix-generated TLS settings into the config, but it was *not* merging arrays. ~~I have eliminated the use of jq and instead did the config merge in Nix to maximize maintainability (since neither of us know enough jq to do it acceptably elegantly).~~ This PR applies the jq hack as proposed in the comments.

###### Motivation for this change
Regression (ish) from #97217 cc @sephii 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions -- via nixops, and it's making configs which appear to have the correct settings
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) -- my computer doesn't have kvm :(
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sephii** [comment]
Thanks for the fix! You can see in the PR that [it originally used another way of merging the JSON](https://github.com/NixOS/nixpkgs/pull/97217#issuecomment-688658299), which caused issues because the way the merge was done didn't create a derivation, which is the reason why `jq` was used in the end.

I checked your PR and the problem seems to be back, since running `nix-build -A nixos/tests/caddy.nix` fails like before.

**Comment by lf-** [comment]
That's super unfortunate.

I'd appreciate if someone who knows more jq can help fix this "properly". @xfix possibly?

**Comment by lf-** [comment]
Can confirm that the gigantic jq command from https://github.com/NixOS/nixpkgs/pull/97217#issuecomment-689647988 *does* fix it as well.

I'm very interested in what about nix-generated JSON causes it to lose the context though, that seems like a bug? Or at least, it seems rather problematic.

**Comment by worldofpeace** [comment]
This needs a rebase.

**Comment by lf-** [comment]
@worldofpeace done

**Comment by lf-** [comment]
So this is the "hack" version proposed in a PR comment elsewhere, as it, I think, passes the tests.

File: nixos/modules/services/web-servers/caddy.nix [hunk 1]
```diff
@@ -21,7 +21,7 @@ let
   '';
   tlsJSON = pkgs.writeText "tls.json" (builtins.toJSON tlsConfig);
   configJSON = pkgs.runCommand "caddy-config.json" { } ''
-    ${pkgs.jq}/bin/jq -s '.[0] * .[1]' ${adaptedConfig} ${tlsJSON} > $out
+    ${pkgs.jq}/bin/jq -s '.[0] * {"apps": {"tls": {"automation": {"policies": (if .[0].apps.tls.automation.policies == .[1]?.apps.tls.automation.policies then .[0].apps.tls.automation.policies else (.[0].apps.tls.automation.policies + .[1]?.apps.tls.automation.policies) end)}}}}' ${adaptedConfig} ${tlsJSON} > $out
```
Comment (worldofpeace) [inline nixos/modules/services/web-servers/caddy.nix:24]: can we put this in a let expression with a comment? (that being the complicated part of the eq expression

Comment (lf-) [reply nixos/modules/services/web-servers/caddy.nix]: Done and reformatted much better :)

**Comment by worldofpeace** [comment]
@ofborg test caddy

**Comment by worldofpeace** [comment]
@lf- thanks for waiting and mentioning this :sparkles: . I think at the time the release was really eating my free time and nixos time. Good to see it's merged finally

**Pull request merged**